File: agent-service/src/agents/review/diff-parser.ts
/**
 * Diff parser — converts raw `git diff` output into a structured representation
 * that the ReviewAgent can pass to the LLM as clean, focused context.
 */

export interface DiffHunk {
  oldStart: number;
  oldLines: number;
  newStart: number;
  newLines: number;
  content: string; // raw hunk text including the @@ header
}

export interface FileDiff {
  path: string;
  oldPath: string | null; // non-null only for renames
  status: "added" | "modified" | "deleted" | "renamed";
  hunks: DiffHunk[];
  /** Flat text of all hunks — easier for LLM consumption */
  rawDiff: string;
}

export interface ParsedDiff {
  files: FileDiff[];
  /** Total lines added across all files */
  totalAdded: number;
  /** Total lines removed across all files */
  totalRemoved: number;
}

const FILE_HEADER_RE = /^diff --git a\/(.*?) b\/(.*)$/;
const HUNK_HEADER_RE = /^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@/;

/**
 * Parse the output of `git diff` (or `git diff HEAD~1`) into structured FileDiffs.
 */
export function parseDiff(rawDiff: string): ParsedDiff {
  const lines = rawDiff.split("\n");
  const files: FileDiff[] = [];
  let current: FileDiff | null = null;
  let currentHunk: DiffHunk | null = null;
  let totalAdded = 0;
  let totalRemoved = 0;

  const pushHunk = () => {
    if (current && currentHunk) {
      current.hunks.push(currentHunk);
      current.rawDiff += currentHunk.content + "\n";
    }
  };

  for (const line of lines) {
    const fileMatch = FILE_HEADER_RE.exec(line);
    if (fileMatch) {
      pushHunk();
      currentHunk = null;
      if (current) files.push(current);
      current = {
        path: fileMatch[2],
        oldPath: fileMatch[1] !== fileMatch[2] ? fileMatch[1] : null,
        status: "modified",
        hunks: [],
        rawDiff: "",
      };
      continue;
    }

    if (!current) continue;

    // Detect file status from git diff headers
    if (line.startsWith("new file mode")) { current.status = "added"; continue; }
    if (line.startsWith("deleted file mode")) { current.status = "deleted"; continue; }
    if (line.startsWith("rename from")) { current.status = "renamed"; continue; }

    const hunkMatch = HUNK_HEADER_RE.exec(line);
    if (hunkMatch) {
      pushHunk();
      currentHunk = {
        oldStart: parseInt(hunkMatch[1], 10),
        oldLines: parseInt(hunkMatch[2] ?? "1", 10),
        newStart: parseInt(hunkMatch[3], 10),
        newLines: parseInt(hunkMatch[4] ?? "1", 10),
        content: line + "\n",
      };
      continue;
    }

    if (currentHunk) {
      currentHunk.content += line + "\n";
      if (line.startsWith("+") && !line.startsWith("+++")) totalAdded++;
      if (line.startsWith("-") && !line.startsWith("---")) totalRemoved++;
    }
  }

  pushHunk();
  if (current) files.push(current);

  return { files, totalAdded, totalRemoved };
}

/**
 * Format a ParsedDiff into a compact LLM-friendly string, optionally
 * truncating very large diffs to stay within token budgets.
 */
export function formatDiffForLLM(diff: ParsedDiff, maxCharsPerFile = 4000): string {
  const parts: string[] = [
    `Changed files: ${diff.files.length} | +${diff.totalAdded} -${diff.totalRemoved}\n`,
  ];

  for (const file of diff.files) {
    const header = `\n--- ${file.status.toUpperCase()}: ${file.path} ---\n`;
    const body = file.rawDiff.length > maxCharsPerFile
      ? file.rawDiff.slice(0, maxCharsPerFile) + "\n... (truncated)"
      : file.rawDiff;
    parts.push(header + body);
  }

  return parts.join("");
}

File: agent-service/src/agent_service/agents/review/diff_parser.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import List, Literal, Optional


@dataclass
class DiffHunk:
    old_start: int
    old_lines: int
    new_start: int
    new_lines: int
    content: str


@dataclass
class FileDiff:
    path: str
    old_path: Optional[str]
    status: Literal["added", "modified", "deleted", "renamed"]
    hunks: List[DiffHunk] = field(default_factory=list)
    raw_diff: str = ""


@dataclass
class ParsedDiff:
    files: List[FileDiff]
    total_added: int
    total_removed: int


_FILE_HEADER_RE = re.compile(r"^diff --git a/(.*?) b/(.*)$")
_HUNK_HEADER_RE = re.compile(r"^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@")


def parse_diff(raw_diff: str) -> ParsedDiff:
    """Parse the output of ``git diff`` into structured FileDiffs."""
    lines = raw_diff.split("\n")
    files: List[FileDiff] = []
    current: Optional[FileDiff] = None
    current_hunk: Optional[DiffHunk] = None
    total_added = 0
    total_removed = 0

    def push_hunk() -> None:
        if current and current_hunk:
            current.hunks.append(current_hunk)
            current.raw_diff += current_hunk.content + "\n"

    for line in lines:
        file_match = _FILE_HEADER_RE.match(line)
        if file_match:
            push_hunk()
            current_hunk = None
            if current:
                files.append(current)
            a_path = file_match.group(1)
            b_path = file_match.group(2)
            current = FileDiff(
                path=b_path,
                old_path=a_path if a_path != b_path else None,
                status="modified",
            )
            continue

        if current is None:
            continue

        if line.startswith("new file mode"):
            current.status = "added"
            continue
        if line.startswith("deleted file mode"):
            current.status = "deleted"
            continue
        if line.startswith("rename from"):
            current.status = "renamed"
            continue

        hunk_match = _HUNK_HEADER_RE.match(line)
        if hunk_match:
            push_hunk()
            current_hunk = DiffHunk(
                old_start=int(hunk_match.group(1)),
                old_lines=int(hunk_match.group(2) or "1"),
                new_start=int(hunk_match.group(3)),
                new_lines=int(hunk_match.group(4) or "1"),
                content=line + "\n",
            )
            continue

        if current_hunk:
            current_hunk.content += line + "\n"
            if line.startswith("+") and not line.startswith("+++"):
                total_added += 1
            if line.startswith("-") and not line.startswith("---"):
                total_removed += 1

    push_hunk()
    if current:
        files.append(current)

    return ParsedDiff(files=files, total_added=total_added, total_removed=total_removed)


def format_diff_for_llm(diff: ParsedDiff, max_chars_per_file: int = 4000) -> str:
    """Format a ParsedDiff into a compact LLM-friendly string."""
    parts = [
        f"Changed files: {len(diff.files)} | +{diff.total_added} -{diff.total_removed}\n"
    ]
    for file in diff.files:
        header = f"\n--- {file.status.upper()}: {file.path} ---\n"
        body = (
            file.raw_diff[:max_chars_per_file] + "\n... (truncated)"
            if len(file.raw_diff) > max_chars_per_file
            else file.raw_diff
        )
        parts.append(header + body)
    return "".join(parts)
